### PR TITLE
Add API for a custom list formatter

### PR DESCRIPTION
**Pull request by mk-pmb**
With this feature I hope I'll be able to format my dumps with custom indentation, maxArrayLength and breakLength.
<del>Edit: Nah, put it on hold, I'll implement an even better version tomorrow.</del> Edit: done.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -8,28 +8,35 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
...
     if (depth >= maxDepth && maxDepth > 0 && obj && typeof obj === 'object') {
         return '[Object]';
     }
-    
+
     if (seen === undefined) seen = [];
     else if (indexOf(seen, obj) >= 0) {
         return '[Circular]';
     }
-
```
Comment (ljharb) [inline index.js]: please revert all unrelated whitespace changes.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "object-inspect",
-  "version": "1.2.1",
+  "version": "1.2.2",
```
Comment (ljharb) [inline package.json]: Version numbers must *never* be bumped in PRs - please revert this.

**Comment by ljharb** [comment]
I'm not sure I understand the use case - but regardless, the tests (that don't have your option provided) must not change, otherwise it's a breaking change.

**Comment by mk-pmb** [comment]
I've reverted the version bump as requested, and will move the whitespace cleanup to its own PR.

**Comment by mk-pmb** [comment]
The new version has the better API that I announced I'd implement today: Type and size are now separate. I managed to even replicate all the old oddities so tests don't have to change and semver jump could probably be at just patch.

**Comment by mk-pmb** [comment]
New version with the EOL whitespace preserved as per #11.

File: index.js [hunk 3]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
...
+  // size = false: parts.length could be considered as the size but it's
+  //        customary to not display it for this type.
+
+  switch (objType || null) {
+  // We could save lots of cases and achieve more consistency in formating,
+  // but that would be a breaking change so let's defer that for the next
+  // breaking semver jump.
+  case null:
+    break;
+  case 'Array':
```
Comment (ljharb) [inline index.js]: `case` statements should be indented one level farther than the `switch` statement they belong to.

File: index.js [hunk 4]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
...
+    parens = '[]';
+    padType = objType = '';
```
Comment (ljharb) [inline index.js]: Please avoid chaining variable assignments; it harms readability.

File: index.js [hunk 5]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
 var setForEach = hasSet && Set.prototype.forEach;
 var booleanValueOf = Boolean.prototype.valueOf;
 
+function listFmtDefault(objType, size, parts, details) {
+  var parens = '{}', padType = ' ', padParts = ' ';
+  // size = null: The concept of size doesn't easily apply for this type.
+  // size = false: parts.length could be considered as the size but it's
+  //        customary to not display it for this type.
+
+  switch (objType || null) {
```
Comment (ljharb) [inline index.js]: could this be done with if/else, or a map lookup, instead of a switch statement?

File: index.js [hunk 6]
```diff
@@ -61,7 +104,8 @@ module.exports = function inspect_ (obj, opts, depth, seen) {
         for (var i = 0; i < obj.length; i++) {
             xs[i] = has(obj, i) ? inspect(obj[i], obj) : '';
         }
-        return '[ ' + xs.join(', ') + ' ]';
+        return listFmt('Array', false, xs,
+          { depth: depth, opts: opts, obj: obj });
```
Comment (ljharb) [inline index.js]: either function invocations should be entirely on one line, or, `(`, each argument, and `)` should be on separate lines.

File: index.js [hunk 7]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
...
+    break;
+  case 'Array':
+    parens = '[]';
+    padType = objType = '';
+    break;
+  case 'Object':
+    padType = objType = '';
+    break;
+  case 'Error':
+    objType = '[' + details.obj + ']';  // NB: implicit String() conversion
```
Comment (ljharb) [inline index.js]: please use `String(x)` instead of `+ x`, as the latter calls `valueOf` over `toString` (then you'll also be able to remove the comment)

File: index.js [hunk 8]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
...
+    if (!parts.length) {
+      parens = '';
+      padType = '';
+    }
+    break;
+  case 'Map':
+  case 'Set':
+    objType += ' (' + size + ') {';
+    size = null;
+    parens = [ '', '}' ];
```
Comment (ljharb) [inline index.js]: array brackets should not have padding spaces inside them.

File: index.js [hunk 9]
```diff
@@ -61,7 +104,8 @@ module.exports = function inspect_ (obj, opts, depth, seen) {
         for (var i = 0; i < obj.length; i++) {
             xs[i] = has(obj, i) ? inspect(obj[i], obj) : '';
         }
-        return '[ ' + xs.join(', ') + ' ]';
+        return listFmt('Array', false, xs,
```
Comment (ljharb) [inline index.js]: why is the "size" param sometimes `false` and sometimes `null`?

I'd kind of expect arrays to pass their length here, even if the "objType" chose to omit it (I do see your comment in the function above)

File: index.js [hunk 10]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
 var setForEach = hasSet && Set.prototype.forEach;
 var booleanValueOf = Boolean.prototype.valueOf;
 
+function listFmtDefault(objType, size, parts, details) {
```
Comment (ljharb) [inline index.js]: Let's `s/Fmt/Format/g` - more letters don't cost extra.

File: index.js [hunk 11]
```diff
@@ -8,10 +8,53 @@ var setSize = hasSet && setSizeDescriptor && typeof setSizeDescriptor.get === 'f
...
+  return (parens[0] || '') + padType + objType +
+    (padParts || padType) + parts + padParts +
+    (parens[1] || '');
+}
+
 module.exports = function inspect_ (obj, opts, depth, seen) {
     if (!opts) opts = {};
     
     var maxDepth = opts.depth === undefined ? 5 : opts.depth;
+    var listFmt = (opts.listFmt || listFmtDefault);
```
Comment (ljharb) [inline index.js]: I think this option name might make more sense as "listFormatter" or "formatList"?

File: index.js (hunk 7 above)
Comment (mk-pmb) [inline index.js]: Then current master should be fixed independent of this PR, because I took the original expression from [index.js#L79](https://github.com/substack/object-inspect/blob/master/index.js#L79) in order to have least impact on compatibility.

Comment (ljharb) [inline index.js]: Sounds good, thanks for clarifying :-)

**Comment by mk-pmb** [comment]
Thanks for your comments! Maybe someone might fix them and see whether it might then land in master. My own motivation for sharing my new feature is used up though, so I'll just fork it and use it as I see fit.

File: index.js (hunk 7 above)
Comment (ljharb) [inline index.js]: Updated; rebase at your leisure.

**Comment by ljharb** [comment]
@mk-pmb adding a new feature requires eternal maintenance, so if you don't have the energy/motivation to make it through a mild code review, it's pretty clear that you won't be around long term to help maintain it either, which generally means it's a bad idea to add the feature.

Thanks for the contribution - if you change your mind, I'm happy to reopen this.

**Comment by mk-pmb** [comment]
You're right, the maintainance perspective is another good argument to go separate routes from here, as I won't re-configure my editor to preserve EOL space any time soon. ;-)
